src/foam/u2/wizard/StepWizardController.js

```diff
@@ -0,0 +1,321 @@
+        }
+        return currentSection$isValid;
+      }
+    },
+    {
+      name: 'submitted',
+      class: 'Boolean'
+    }
+  ],
+
+  methods: [
+    function setupWizardletListeners(wizardlets) {
+      console.log('step wizard', this, wizardlets);
+      if ( this.wsub ) this.wsub.detach();
+      this.wsub = this.FObject.create();
+
+      var wi = 0, si = 0;
+      wizardlets.forEach((w, wizardletIndex) => {
+
+        // Bind availability listener for wizardlet availability
+        var isAvailable$ = w.isAvailable$;
+        console.log(wizardletIndex, w, w.isAvailable$, isAvailable$);
+        this.wsub.onDetach(isAvailable$.sub(() => {
+          console.log('X');
+          this.onWizardletAvailability(wizardletIndex, isAvailable$.get());
+        }));
+
+        // Bind availability listener for each wizardlet section
+        w.sections.forEach((section, sectionIndex) => {
+          var sectionPosition = this.WizardPosition.create({
+            wizardletIndex: wizardletIndex,
+            sectionIndex: sectionIndex,
+          });
+          var isAvailable$ = section.isAvailable$;
+          this.wsub.onDetach(isAvailable$.sub(() => {
+            console.log('Y');
+            this.onWizardPositionAvailabilityUpdate(
+              sectionPosition, isAvailable$.get());
+          }));
+        });
+
+      });
+
+    },
+    function detach() {
+      this.wsub.detach();
+      this.SUPER();
+    },
+    function positionAfter(pos) {
+      let subWi = pos.wizardletIndex
+      let subSi = pos.sectionIndex;
+      if ( subSi >= this.wizardlets[subWi].sections.length - 1 ) {
+        if ( subWi >= this.wizardlets.length - 1 ) return null;
+        subSi = 0;
+        subWi++;
+      } else {
+        subSi++;
+      }
+      return this.WizardPosition.create({
+        wizardletIndex: subWi,
+        sectionIndex: subSi,
+      });
+    },
+    function positionBefore(pos) {
+      let subWi = pos.wizardletIndex;
+      let subSi = pos.sectionIndex;
+      if ( subSi == 0 ) {
+        if ( subWi == 0 ) return null;
+        subWi--;
+        // Skip past steps with no sections
+        while ( this.wizardlets[subWi].sections.length < 1 ) subWi--;
+        if ( subWi < 0 ) return null;
+        subSi = this.wizardlets[subWi].sections.length - 1;
+      } else {
+        subSi--;
+      }
+      return this.WizardPosition.create({
+        wizardletIndex: subWi,
+        sectionIndex: subSi,
+      });
+    },
+    function nextAvailable(pos, iter) {
+      for ( let p = iter(pos) ; p != null ; p = iter(p) ) {
+        let wizardlet = this.wizardlets[p.wizardletIndex]
+        if ( ! wizardlet.isVisible ) continue;
+
+        if (
+          wizardlet.sections.length > 0 &&
+          wizardlet.sections[p.sectionIndex].isAvailable
+        ) return p;
+      }
+
+      return null;
+    },
+    function saveProgress() {
+      return this.wizardlets.reduce((p, wizardlet) =>
```

Comment:
If we have visited a wizardlet - why are we doing things differently? Should we not save it again?

It would be nice to not try to force synchronization of js code with reduce and promises all over the place.

Reply:
`saveProgress` isn't actually called. I copied it from the other wizard and updated it - now I'm just keeping there on the short term in case we need it.

On your second point, if we don't save in order the saves don't happen in order, which I've been told on multiple occasions needs to happen. (I can see the rationale - it increases client time but saves on server load)